service/src/main/java/bio/terra/externalcreds/services/OAuth2Service.java

```diff
@@ -0,0 +1,141 @@
+        .getAuthorizationRequestUri();
+  }
+
+  private OAuth2AuthorizationRequest createOAuth2AuthorizationRequest(
+      String redirectUri,
+      Set<String> scopes,
+      String state,
+      ClientRegistration providerClient,
+      Map<String, Object> additionalAuthorizationParameters) {
+
+    return OAuth2AuthorizationRequest.authorizationCode()
+        .authorizationUri(providerClient.getProviderDetails().getAuthorizationUri())
+        .redirectUri(redirectUri)
+        .clientId(providerClient.getClientId())
+        .scopes(scopes)
+        .state(state)
+        .additionalParameters(additionalAuthorizationParameters)
+        .build();
+  }
+
+  /**
+   * After authentication, the resulting code should be used here
+   *
+   * @param providerClient identity provider client, see {@link ProviderClientCache}
+   * @param redirectUri uri the user will be directed to after authentication
+   * @param scopes scopes requested for authentication
+   * @param state oauth thing
+   * @param additionalAuthorizationParameters any other parameters
+   * @param authorizationCode code resulting from authentication
+   * @return token response containing access and refresh tokens
+   */
+  public OAuth2AccessTokenResponse authorizationCodeExchange(
+      ClientRegistration providerClient,
+      String authorizationCode,
+      String redirectUri,
+      Set<String> scopes,
+      String state,
+      Map<String, Object> additionalAuthorizationParameters) {
+
+    OAuth2AuthorizationRequest authRequest =
+        createOAuth2AuthorizationRequest(
+            redirectUri, scopes, state, providerClient, additionalAuthorizationParameters);
+
+    OAuth2AuthorizationResponse authResponse =
+        OAuth2AuthorizationResponse.success(authorizationCode)
+            .redirectUri(redirectUri)
+            .state(state)
+            .build();
+
+    DefaultAuthorizationCodeTokenResponseClient tokenResponseClient =
+        new DefaultAuthorizationCodeTokenResponseClient();
+    OAuth2AuthorizationCodeGrantRequest codeGrantRequest =
+        new OAuth2AuthorizationCodeGrantRequest(
+            providerClient, new OAuth2AuthorizationExchange(authRequest, authResponse));
+
+    return tokenResponseClient.getTokenResponse(codeGrantRequest);
+  }
+
+  /**
+   * Given an refresh token, get an access token
+   *
+   * @param providerClient identity provider client, see {@link ProviderClientCache}
+   * @param refreshToken
+   * @return token response containing access and refresh tokens, note that if there is a refresh
+   *     token in this response it should replace the original refresh token which is likely invalid
+   */
+  public OAuth2AccessTokenResponse authorizeWithRefreshToken(
+      ClientRegistration providerClient, OAuth2RefreshToken refreshToken) {
+    // the OAuth2RefreshTokenGrantRequest requires an access token to be specified but not a valid one
```

Comment:
the linter breaks up this line, could you just run it and then commit so we don't wind up with churn?